Tests/ApolloCodegenTests/CodeGenAST/IROperationBuilderTests.swift

```diff
@@ -0,0 +1,3031 @@
+import XCTest
+import Nimble
+import OrderedCollections
+@testable import ApolloCodegenLib
+import ApolloTestSupport
+import ApolloCodegenTestSupport
+import ApolloAPI
+
+class IROperationBuilderTests: XCTestCase {
+
+  var mockCompilationResult: CompilationResult!
+  var operation: CompilationResult.OperationDefinition!
+  var subject: IR.Operation!
+
+  override func setUp() {
+    super.setUp()
+    mockCompilationResult = CompilationResult.mock()
+    operation = CompilationResult.OperationDefinition.mock()
+  }
+
+  override func tearDown() {
+    mockCompilationResult = nil
+    operation = nil
+    subject = nil
+    super.tearDown()
+  }
+
+  // MARK: = Helpers
+
+  func buildSubjectOperation() {
+    subject = IR(compilationResult: mockCompilationResult).build(operation: operation)
+  }
+
+  // MARK: - Children Computation
+
+  // MARK: Children - Fragment Type
+
+  /// Example:
+  /// query {
+  ///  allAnimals {
+  ///    ...AnimalDetails
+  ///  }
+  /// }
+  ///
+  /// fragment AnimalDetails on Animal {
+  ///   species
+  /// }
+  /// Expected:
+  /// Animal.children should not include a type case for asAnimal
+  func test__children__initWithNamedFragmentOnTheSameType_hasNoChildTypeCase() {
+    // given
+    let Interface_Animal = GraphQLInterfaceType.mock("Animal")
+
+    let animalDetails = CompilationResult.FragmentDefinition.mock("AnimalDetails", type: Interface_Animal)
+
+    operation = .mock(selections: [
+      .field(.mock(
+        "allAnimals",
+        selectionSet: .mock(
+          parentType: Interface_Animal,
+          selections: [
+            .fragmentSpread(animalDetails),
+          ]
+        )))])
+
+    // when
+    buildSubjectOperation()
+
+    let allAnimals = self.subject[field: "query"]?[field: "allAnimals"]?.selectionSet
+
+    // then
+    expect(allAnimals?.selections.typeCases).to(beEmpty())
+  }
+
+  /// Example:
```

Comment:
These comments-as-examples are super helpful for test readability. I think we can improve them slightly though:
* include relevant SDL for the operation
* enclose in "```" top and bottom so they're more readable in Xcode

This one would end up looking like..
```
  /// ```
  /// SDL:
  /// type Query {
  ///   allAnimals: [Animal!]!
  /// }
  /// interface Animal {
  ///   species: String!
  /// }
  /// type Bird implements Animal {
  ///   species: String!
  /// }
  ///
  /// OPERATION:
  /// query {
  ///  allAnimals {
  ///    ...BirdDetails
  /// }
  ///
  /// fragment BirdDetails on Bird {
  ///   species
  /// }
  /// Expected:
  /// AllAnimals.children: [AsBird]
  /// ```
```